### ISSUE TITLE
Lower mIoU

### ISSUE THREAD

**Issue by mehditlili** [Tramac/Fast-SCNN-pytorch#2]
You report 54% mIoU, I got around 49% after training with the train script (160 epochs, batch size 12). Any idea why this is much lower as the reported 69% in the paper?

**Comment by Tramac** [comment]
I think the crop size maybe the main reason，the image size of paper is 2048*1024，768*768 in mine.

**Comment by Tramac** [comment]
Hi, you can try changing SGD optimizer to Adam, this will have a big improvement when the epochs is not enough.

**Comment by mehditlili** [comment]
Okay thanks, I will try it out.
also I am struggling to understand this part
`    def __init__(self, root='./datasets/citys', split='train', mode=None, transform=None,
                 base_size=520, crop_size=480, **kwargs):
        super(CitySegmentation, self).__init__()
        self.root = root
        self.split = split
        self.mode = mode if mode is not None else split
        self.transform = transform
        self.base_size = base_size
        self.crop_size = crop_size
        self.images, self.mask_paths = _get_city_pairs(self.root, self.split)
        assert (len(self.images) == len(self.mask_paths))
        if len(self.images) == 0:
            raise RuntimeError("Found 0 images in subfolders of: " + self.root + "\n")
        self.valid_classes = [7, 8, 11, 12, 13, 17, 19, 20, 21, 22,
                              23, 24, 25, 26, 27, 28, 31, 32, 33]
        self._key = np.array([-1, -1, -1, -1, -1, -1,
                              -1, -1, 0, 1, -1, -1,
                              2, 3, 4, -1, -1, -1,
                              5, -1, 6, 7, 8, 9,
                              10, 11, 12, 13, 14, 15,
                              -1, -1, 16, 17, 18])
        self._mapping = np.array(range(-1, len(self._key) - 1)).astype('int32')`
What If I want to train to recognize omy own class with id 100, what are self._mapping and self._key used for?

**Comment by Tramac** [comment]
Hi, the cityscapes dataset is special.
- The actual number of categories is 35(the length of `self._key`), where the number of valid categories is 19 and invalid categories is 16.  
- The index of valid categories is `self.valid_classes`. 
- `self._mapping` is a numpy array, ranging from -1 to `classes-1`(For cityscapes: -1~34).
- `trainId` in cityscapes: `np.array([255, 255, 255, 255, 255, 255, 255, 0, 1, 255, 255, 2, 3, 4, 255, 255, 255, 5, 255, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 255, 255, 16, 17, 18, -1])`, the invalid value `255` is set `-1`.
- `self._key` map the actual category (in `self.valid_classes`) to the desired category(-1~18).

But, the function `_class_to_index()` may be unnecessary for your own dataset if the labeled value is continuous(0~99 for yours).
If there is something unclear, please continue to post the question.

**Comment by mehditlili** [comment]
Okay thank you, just removed the whole thing and trained direclty with the values in my label images (0 and 1)

**Comment by jiaobingle** [comment]
@Tramac 
Hi, the labeled value in my dataset is 0,1,2,3 and ignore_label=255, and i modified:
        self.valid_classes = [7, 8, 11, 12, 13, 17, 19, 20, 21, 22,
                              23, 24, 25, 26, 27, 28, 31, 32, 33]
        self._key = np.array([-1, -1, -1, -1, -1, -1,
                              -1, -1, 0, 1, -1, -1,
                              2, 3, 4, -1, -1, -1,
                              5, -1, 6, 7, 8, 9,
                              10, 11, 12, 13, 14, 15,
                              -1, -1, 16, 17, 18])
        self._mapping = np.array(range(-1, len(self._key) - 1)).astype('int32')

as: 
        self.valid_classes = [0,1,2,3]    #19  
        self._key = np.array([0,1,2,3,255])  #35   ignore_label=255
        self._mapping = (self._key).astype('int32')

but i also did not solve it, because the invalid value 255 has not been setted -1, and how should i do next, thx!
(in my dataset ,i can not set the invalid value 255 to -1, so "self._mapping = np.array(range(-1, len(self._key) - 1)).astype('int32')" does not work)

**Comment by Tramac** [comment]
Hi, if your labeled value is 0~3, the function [`_class_to_index`](https://github.com/Tramac/Fast-SCNN-pytorch/blob/4d53b39d4704c0b7663b5e68332d22f8ad3eda41/data_loader/cityscapes.py#L156) in [`_mask_transform`](https://github.com/Tramac/Fast-SCNN-pytorch/blob/4d53b39d4704c0b7663b5e68332d22f8ad3eda41/data_loader/cityscapes.py#L155) is no necessary. You can remove this function directly.
For your question, you need to change function [`_mask_transform`](https://github.com/Tramac/Fast-SCNN-pytorch/blob/4d53b39d4704c0b7663b5e68332d22f8ad3eda41/data_loader/cityscapes.py#L155) as follows:
```
def _mask_transform(self, mask):
    target = np.array(mask).astype('int32')
    target[target == 255] = -1
    return torch.from_numpy(target).long()
```
More dataloader can see my another [project](https://github.com/Tramac/Awesome-semantic-segmentation-pytorch/tree/master/core/data/dataloader)

**Comment by jiaobingle** [comment]
@Tramac  thanks, solved it successfully.

**Comment by lxtGH** [comment]
＂We found that the performance of DCNNs can be improved by training for higher number
of iterations, hence we train our model for 1,000 epochs unless otherwise stated, using the
Cityescapes dataset＂ Hi , did you train longer than this ? @Tramac
